src/librustc_front/visit.rs

```diff
@@ -203,26 +203,31 @@ pub fn walk_item<'v, V: Visitor<'v>>(visitor: &mut V, item: &'v Item) {
         ItemExternCrate(..) => {}
         ItemUse(ref vp) => {
             match vp.node {
-                ViewPathSimple(ident, ref path) => {
-                    visitor.visit_ident(vp.span, ident);
+                ViewPathSimple(_ident, ref path) => {
                     visitor.visit_path(path, item.id);
                 }
                 ViewPathGlob(ref path) => {
                     visitor.visit_path(path, item.id);
                 }
                 ViewPathList(ref prefix, ref list) => {
-                    for id in list {
-                        match id.node {
-                            PathListIdent { name, .. } => {
-                                visitor.visit_ident(id.span, name);
+                    let mut path = prefix.clone();
+                    for item in list {
+                        match item.node {
+                            PathListIdent { name, id, .. } => {
+                                path.span = item.span;
+                                path.segments.push(PathSegment {
+                                    identifier: name,
+                                    parameters: PathParameters::none()
+                                });
+                                visitor.visit_path(unsafe{&*(&path as *const _)}, id);
```

Comment:
Yes, the prefix is cloned and components of the prefix are visited repeatedly. I supposed this couldn't be done optimally without introducing a new `visit_xxx` method, but introducing it seemed.. intrusive and too far away from stability.rs. I'll surely do it if it's acceptable.

I also had an idea to desugar `use a::{b, c};` into `use a::b; use a::b;` during lowering to HIR, it would avoid repeated cloning of the prefix, but checks of the prefix components would still be duplicated.